// @netlify/zip-it-and-ship-it/dist/runtimes/node/bundlers/none/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSrcFiles = void 0;
const path_1 = require("path");
const error_js_1 = require("../../../../utils/error.js");
const base_path_js_1 = require("../../utils/base_path.js");
const included_files_js_1 = require("../../utils/included_files.js");
const node_version_js_1 = require("../../utils/node_version.js");
const package_json_js_1 = require("../../utils/package_json.js");
/**
 * This bundler is a simple no-op bundler, that does no bundling at all.
 * It returns the detected moduleFormat and the mainFile + includedFiles from the config.
 */
// Uses the same algorithm as the ESM_FILE_FORMAT resolver from Node.js to get
// the right module format for a function based on its main file.
// See https://nodejs.org/api/esm.html#resolver-algorithm-specification.
const getModuleFormat = async function (mainFile) {
    const extension = (0, path_1.extname)(mainFile);
    if (extension === '.mjs') {
        return "esm" /* ModuleFormat.ESM */;
    }
    if (extension === '.cjs') {
        return "cjs" /* ModuleFormat.COMMONJS */;
    }
    const packageJson = await (0, package_json_js_1.getPackageJsonIfAvailable)((0, path_1.dirname)(mainFile));
    if (packageJson.type === 'module') {
        return "esm" /* ModuleFormat.ESM */;
    }
    return "cjs" /* ModuleFormat.COMMONJS */;
};
const getSrcFiles = async function ({ config, mainFile }) {
    const { includedFiles = [], includedFilesBasePath } = config;
    const { excludePatterns, paths: includedFilePaths } = await (0, included_files_js_1.getPathsOfIncludedFiles)(includedFiles, includedFilesBasePath);
    const includedPaths = (0, included_files_js_1.filterExcludedPaths)(includedFilePaths, excludePatterns);
    return { srcFiles: [mainFile, ...includedPaths], includedFiles: includedPaths };
};
exports.getSrcFiles = getSrcFiles;
const bundle = async ({ basePath, config, extension, featureFlags, filename, mainFile, name, pluginsModulesPath, runtime, srcDir, srcPath, stat, }) => {
    const { srcFiles, includedFiles } = await (0, exports.getSrcFiles)({
        basePath,
        config: {
            ...config,
            includedFilesBasePath: config.includedFilesBasePath || basePath,
        },
        extension,
        featureFlags,
        filename,
        mainFile,
        name,
        pluginsModulesPath,
        runtime,
        srcDir,
        srcPath,
        stat,
    });
    const dirnames = srcFiles.map((filePath) => (0, path_1.normalize)((0, path_1.dirname)(filePath)));
    const moduleFormat = await getModuleFormat(mainFile);
    const nodeSupport = (0, node_version_js_1.getNodeSupportMatrix)(config.nodeVersion);
    if (moduleFormat === "esm" /* ModuleFormat.ESM */ && !nodeSupport.esm) {
        throw new error_js_1.FunctionBundlingUserError(`Function file is an ES module, which the Node.js version specified in the config (${config.nodeVersion}) does not support. ES modules are supported as of version 14 of Node.js.`, {
            functionName: name,
            runtime: "js" /* RuntimeType.JAVASCRIPT */,
            bundler: "none" /* NodeBundlerType.NONE */,
        });
    }
    return {
        basePath: (0, base_path_js_1.getBasePath)(dirnames),
        includedFiles,
        inputs: srcFiles,
        mainFile,
        moduleFormat,
        srcFiles,
    };
};
const bundler = { bundle, getSrcFiles: exports.getSrcFiles };
exports.default = bundler;
